[metal] `MTLDevice` is thread-safe

uael opened the pull request:
**Description**
`MTLDevice` **look like** is safe to access from different threads, this remove the `Mutex` around it. Feel free to close if it's actually not, the Apple documentation is very unclear about that and only notice about the [thread safety of the command queue](https://developer.apple.com/documentation/metal/mtlcommandqueue)

**Testing**
Tested manually on iOS.

**Checklist**

- [x] Run `cargo fmt`.
- [x] ~~Run `taplo format`.~~
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [x] ~~`--target wasm32-unknown-unknown`~~
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

uael commented on wgpu-hal/src/metal/device.rs:
```diff
@@ -362,7 +359,7 @@ impl super::Device {
     }
 
-    pub fn raw_device(&self) -> &Mutex<metal::Device> {
+    pub fn raw_device(&self) -> &metal::Device {
```
Add an entry to `CHANGELOG.md` if we are going to merge this

madsmtm reviewed (approved):
I agree that this is thread-safe - you can already safely call `MTLCopyAllDevices()` and/or `MTLCreateSystemDefaultDevice()` from any thread, and access the same instances that way.

cwfitzgerald commented:
Alright, this would be good to have this, but there are CI issues.

madsmtm commented:
For the record, Apple has in Xcode 26 marked the following types as `@Sendable` in Swift (and `objc2-metal` marks them as `Send + Sync` in https://github.com/madsmtm/objc2/pull/762):
- `MTLArgument`
- `MTLCommandQueue`
- `MTLComputePipelineReflection`
- `MTLComputePipelineState`
- `MTLCounter`
- `MTLCounterSet`
- `MTLDepthStencilState`
- `MTLDevice` <-- this
- `MTLDynamicLibrary`
- `MTLEvent`
- `MTLSharedEventListener`
- `MTLSharedEventHandle`
- `MTLFence`
- `MTLFunctionHandle`
- `MTLIOCommandQueue`
- `MTLIOFileHandle`
- `MTLFunction`
- `MTLFunctionReflection`
- `MTLLibrary`
- `MTLLogState`
- `MTLRasterizationRateMap`
- `MTLRenderPipelineReflection`
- `MTLRenderPipelineState`
- `MTLSamplerState`

uael commented:
Sorry for the late response. I expect to handle this under 2 weeks

cwfitzgerald commented:
Friendly poke :)

uael commented:
Sorry definitely on my mind but I didn't find the time yet

andyleiserson commented:
I am actively working on the switch to objc2-metal, which may make this obsolete.

madsmtm commented:
> I am actively working on the switch to objc2-metal, which may make this obsolete.

No, I haven't touched these in my `objc2-metal` PR, this PR could land either before or after that.

andyleiserson commented:
I see, it will conflict, but the removal of the mutex applies either way.

cwfitzgerald reviewed (approved):
This looks fine, but needs to be rebased and CI fixed.

uael commented:
Rebased and fixed clippy warnings.

I also removed Mutex around `MTLCommandQueue` and `CAMetalLayer` as they appear to be thread safe as well (@madsmtm please confirm this). But it's removing a lot of synchronization points and it might not be a good a idea to do so.

madsmtm commented:
Not so sure about the layer, I haven't found concrete documentation that those are thread-safe IIRC.

The big danger with removing synchronisation is if something elsewhere implicitly relied on it. I'm not one making the decisions here, but I think it would be prudent to start with just the `MTLDevice` that you did originally (I'm more confident that those are 100% thread-safe), and maybe work on the rest individually in a separate PR?

uael commented:
> Not so sure about the layer, I haven't found concrete documentation that those are thread-safe IIRC.
> 
> The big danger with removing synchronisation is if something elsewhere implicitly relied on it. I'm not one making the decisions here, but I think it would be prudent to start with just the `MTLDevice` that you did originally (I'm more confident that those are 100% thread-safe), and maybe work on the rest individually in a separate PR?

Yes that was my concern, let's do `MTLDevice` only for now.

uael committed:
[metal] `MTLDevice` is thread-safe
